### ISSUE TITLE
Rule proposal: Forbid `try-catch` inside `try-catch`

### ISSUE THREAD

**Issue by fisker** [sindresorhus/eslint-plugin-unicorn#2189]
### Description

Prevent mistake, when mean to use `try-finally` in `try-catch`.

### Fail

```
try {
  try {
    doSomeThing()
  } catch { // <- Typo, should be finally
    // Clean up
  } 
} catch {
  // Handle error
}
```

### Pass

```
try {
  try {
    doSomeThing()
  } finally {
    // Clean up
  } 
} catch {
  // Handle error
}
```

### Additional Info

I was thinking proposal this to [`no-useless-catch`](https://eslint.org/docs/latest/rules/no-useless-catch) at first, but in this case the `catch` block not actually "useless", since it could mean to catch error thrown by inner `catch` block.

**Comment by fregante** [comment]
Why does this kick in? 

```js
try {
  try {
    doSomeThing()
  } catch {
    // I agree that this means the outer `catch` is
    // never triggered, so it's broken
  } 
} catch {
  // Handle error
}
```

What if:


```diff
  try {
    try {
      doSomeThing()
    } catch {
      // Cleanup
    } 
+   someOtherThrowyStatement()
  } catch {
    // Handle error
  }
```

Suddenly the inner `catch` is no longer "useless" because it ensures `someOtherThrowyStatement` runs. And then if `someOtherThrowyStatement` throws, there's still the outer `catch`.

Or: 


```diff
  try {
    try {
      doSomeThing()
    } catch {
      // Cleanup
    } 

+   try {
+     doSomeOtherThing()
+   } catch {
+     // Cleanup
+   } 
  } catch {
    // Handle error
  }
```

Still awkward and unnecessary, but the first `catch` is still needed in order to run the second `catch`

Nesting try/catch seems to be an antipattern to begin with, though

**Comment by fisker** [comment]
> Suddenly the inner catch is no longer "useless" because it ensures someOtherThrowyStatement runs. And then if someOtherThrowyStatement throws, there's still the outer catch.

That's why I add "lonely" to the title.
![Screenshot_2023-08-26-19-41-55-405_com microsoft emmx canary-edit](https://github.com/sindresorhus/eslint-plugin-unicorn/assets/172584/d9aed344-8577-462c-925c-e00cd51d306e)


> Note that each of the inner catch clauses could throw their own errors. 

Mentioned in "Additional Info"

> in this case the catch block not actually "useless", since it could mean to catch error thrown by inner catch block.

If we only check cases that the inner `try-catch` is 
the only child in the outer `try` block. Then we should still report it, since the error better be caught in the `catch` block.

---

Anyway with this limitation, we'll have to ignore the following case.


```js
try {
  let foo; // <-- extra node
  try {
    doSomeThing()
  } finally {
    // Clean up
  } 
} catch {
  // Handle error
}
```

**Comment by fregante** [comment]
Ok, so you want to use this as a "cleanup before catch" pattern:

```js
try {
  try {
    doSomeThing()
  } finally {
    // Clean up
  } 
} catch {
  // Handle error
}
```

I've never seen this and I'd probably prefer a linear try/catch/finally in every case. When is it necessary to cleaning up strictly before catching the error?

**Comment by fisker** [comment]
Example, when the `catch` block is a `await expression`

```js
showLoading()

try {
  try {
    await fetch(...)
  } finally {
    hideLoading()
  } 
} catch {
  await showErrorMessage()
  return
}

// ...
```

If we use

```js
showLoading()

try {
  await fetch(...)
} catch {
  await showErrorMessage()
  return
} finally {
  hideLoading()
} 

// ...
```

`hideLoading` will execute after `showErrorMessage`.

**Comment by fregante** [comment]
👍 

JavaScript needs a new statement 😃 

```JavaScript 
showLoading()

try {
  await fetch(...)
} then {
  hideLoading()
} catch {
  await showErrorMessage()
  return
}
```

**Comment by sindresorhus** [comment]
This is now covered by `unicorn/try-complexity`, which reports nested `try` statements inside a `try` block. I don’t think we need a separate narrower rule for this, especially since changing `catch` to `finally` is not semantics-preserving.
